File: go/storage/vector_store.go
package storage

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"math"
	"net/http"

	"github.com/google/uuid"
)

const (
	// TextCollection is the Qdrant collection for text embeddings.
	TextCollection = "memory_text"
	// VisualCollection is the Qdrant collection for visual (CLIP) embeddings.
	VisualCollection = "memory_visual"
)

// VectorStore provides vector similarity search via Qdrant REST API.
type VectorStore struct {
	baseURL    string
	httpClient *http.Client
}

// NewVectorStore creates a new VectorStore targeting the given Qdrant REST API URL.
func NewVectorStore(baseURL string) *VectorStore {
	return &VectorStore{
		baseURL:    baseURL,
		httpClient: &http.Client{},
	}
}

// Initialize ensures the text and visual collections exist in Qdrant.
func (v *VectorStore) Initialize(ctx context.Context, textDim, visualDim int) error {
	if err := v.ensureCollection(ctx, TextCollection, textDim); err != nil {
		return fmt.Errorf("ensuring text collection: %w", err)
	}
	if err := v.ensureCollection(ctx, VisualCollection, visualDim); err != nil {
		return fmt.Errorf("ensuring visual collection: %w", err)
	}
	return nil
}

// Close is a no-op for the REST-based vector store.
func (v *VectorStore) Close() error {
	return nil
}

func (v *VectorStore) ensureCollection(ctx context.Context, name string, dim int) error {
	// Check if collection exists
	resp, err := v.doRequest(ctx, "GET", fmt.Sprintf("/collections/%s", name), nil)
	if err == nil && resp.StatusCode == http.StatusOK {
		resp.Body.Close()
		return nil
	}
	if resp != nil {
		resp.Body.Close()
	}

	// Create collection
	body := map[string]interface{}{
		"vectors": map[string]interface{}{
			"size":     dim,
			"distance": "Cosine",
		},
	}
	resp, err = v.doRequest(ctx, "PUT", fmt.Sprintf("/collections/%s", name), body)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		data, _ := io.ReadAll(resp.Body)
		return fmt.Errorf("failed to create collection %s: %s", name, string(data))
	}
	return nil
}

// ---------- Text embeddings ----------

// UpsertTextVector inserts or updates a text embedding. Returns the point ID.
func (v *VectorStore) UpsertTextVector(
	ctx context.Context,
	memoryID string, vector []float64,
	tier string, valence, arousal float64,
	sessionID, createdAt string,
) (string, error) {
	pointID := uuid.New().String()

	body := map[string]interface{}{
		"points": []map[string]interface{}{
			{
				"id":     pointID,
				"vector": vector,
				"payload": map[string]interface{}{
					"memory_id":  memoryID,
					"tier":       tier,
					"valence":    valence,
					"arousal":    arousal,
					"session_id": sessionID,
					"created_at": createdAt,
				},
			},
		},
	}

	resp, err := v.doRequest(ctx, "PUT", fmt.Sprintf("/collections/%s/points", TextCollection), body)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		data, _ := io.ReadAll(resp.Body)
		return "", fmt.Errorf("upsert text vector failed: %s", string(data))
	}
	return pointID, nil
}

// TextSearchResult holds a result from text vector search.
type TextSearchResult struct {
	MemoryID string  `json:"memory_id"`
	Score    float64 `json:"score"`
	Tier     string  `json:"tier"`
	Valence  float64 `json:"valence"`
	Arousal  float64 `json:"arousal"`
}

// SearchText searches for nearest text embeddings.
func (v *VectorStore) SearchText(ctx context.Context, queryVector []float64, limit int, tierFilter *string) ([]TextSearchResult, error) {
	body := map[string]interface{}{
		"vector":       queryVector,
		"limit":        limit,
		"with_payload": true,
	}

	if tierFilter != nil {
		body["filter"] = map[string]interface{}{
			"must": []map[string]interface{}{
				{
					"key": "tier",
					"match": map[string]interface{}{
						"value": *tierFilter,
					},
				},
			},
		}
	}

	resp, err := v.doRequest(ctx, "POST", fmt.Sprintf("/collections/%s/points/search", TextCollection), body)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	var result struct {
		Result []struct {
			ID      interface{}            `json:"id"`
			Score   float64                `json:"score"`
			Payload map[string]interface{} `json:"payload"`
		} `json:"result"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, fmt.Errorf("decoding search response: %w", err)
	}

	var results []TextSearchResult
	for _, r := range result.Result {
		ts := TextSearchResult{
			Score: r.Score,
		}
		if p := r.Payload; p != nil {
			if v, ok := p["memory_id"].(string); ok {
				ts.MemoryID = v
			}
			if v, ok := p["tier"].(string); ok {
				ts.Tier = v
			}
			if v, ok := p["valence"].(float64); ok {
				ts.Valence = v
			}
			if v, ok := p["arousal"].(float64); ok {
				ts.Arousal = v
			}
		}
		results = append(results, ts)
	}
	return results, nil
}

// ---------- Visual embeddings ----------

// UpsertVisualVector inserts or updates a visual (CLIP) embedding. Returns the point ID.
func (v *VectorStore) UpsertVisualVector(
	ctx context.Context,
	memoryID string, vector []float64,
	sessionID, createdAt string,
) (string, error) {
	pointID := uuid.New().String()

	body := map[string]interface{}{
		"points": []map[string]interface{}{
			{
				"id":     pointID,
				"vector": vector,
				"payload": map[string]interface{}{
					"memory_id":  memoryID,
					"session_id": sessionID,
					"created_at": createdAt,
				},
			},
		},
	}

	resp, err := v.doRequest(ctx, "PUT", fmt.Sprintf("/collections/%s/points", VisualCollection), body)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		data, _ := io.ReadAll(resp.Body)
		return "", fmt.Errorf("upsert visual vector failed: %s", string(data))
	}
	return pointID, nil
}

// VisualSearchResult holds a result from visual vector search.
type VisualSearchResult struct {
	MemoryID string  `json:"memory_id"`
	Score    float64 `json:"score"`
}

// SearchVisual searches for nearest visual embeddings.
func (v *VectorStore) SearchVisual(ctx context.Context, queryVector []float64, limit int) ([]VisualSearchResult, error) {
	body := map[string]interface{}{
		"vector":       queryVector,
		"limit":        limit,
		"with_payload": true,
	}

	resp, err := v.doRequest(ctx, "POST", fmt.Sprintf("/collections/%s/points/search", VisualCollection), body)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	var result struct {
		Result []struct {
			ID      interface{}            `json:"id"`
			Score   float64                `json:"score"`
			Payload map[string]interface{} `json:"payload"`
		} `json:"result"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, fmt.Errorf("decoding visual search response: %w", err)
	}

	var results []VisualSearchResult
	for _, r := range result.Result {
		vs := VisualSearchResult{Score: r.Score}
		if p := r.Payload; p != nil {
			if mid, ok := p["memory_id"].(string); ok {
				vs.MemoryID = mid
			}
		}
		results = append(results, vs)
	}
	return results, nil
}

// Similarity computes cosine similarity between two points in the text collection.
// Returns nil if either point is not found.
func (v *VectorStore) Similarity(ctx context.Context, pointIDA, pointIDB string) (*float64, error) {
	body := map[string]interface{}{
		"ids":          []string{pointIDA, pointIDB},
		"with_vector":  true,
		"with_payload": false,
	}

	resp, err := v.doRequest(ctx, "POST", fmt.Sprintf("/collections/%s/points", TextCollection), body)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	var result struct {
		Result []struct {
			ID     interface{} `json:"id"`
			Vector []float64   `json:"vector"`
		} `json:"result"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, err
	}
	if len(result.Result) < 2 {
		return nil, nil
	}

	sim := cosineSimilarity(result.Result[0].Vector, result.Result[1].Vector)
	return &sim, nil
}

// DeletePoint deletes all points for a given memory_id from a collection.
func (v *VectorStore) DeletePoint(ctx context.Context, collection, memoryID string) error {
	body := map[string]interface{}{
		"filter": map[string]interface{}{
			"must": []map[string]interface{}{
				{
					"key": "memory_id",
					"match": map[string]interface{}{
						"value": memoryID,
					},
				},
			},
		},
	}

	resp, err := v.doRequest(ctx, "POST", fmt.Sprintf("/collections/%s/points/delete", collection), body)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		data, _ := io.ReadAll(resp.Body)
		return fmt.Errorf("delete point failed: %s", string(data))
	}
	return nil
}

// ---------- Helpers ----------

func (v *VectorStore) doRequest(ctx context.Context, method, path string, body interface{}) (*http.Response, error) {
	var bodyReader io.Reader
	if body != nil {
		data, err := json.Marshal(body)
		if err != nil {
			return nil, fmt.Errorf("marshalling request body: %w", err)
		}
		bodyReader = bytes.NewReader(data)
	}

	req, err := http.NewRequestWithContext(ctx, method, v.baseURL+path, bodyReader)
	if err != nil {
		return nil, fmt.Errorf("creating request: %w", err)
	}
	if body != nil {
		req.Header.Set("Content-Type", "application/json")
	}

	return v.httpClient.Do(req)
}

func cosineSimilarity(a, b []float64) float64 {
	if len(a) != len(b) || len(a) == 0 {
		return 0.0
	}
	var dot, normA, normB float64
	for i := range a {
		dot += a[i] * b[i]
		normA += a[i] * a[i]
		normB += b[i] * b[i]
	}
	norm := math.Sqrt(normA) * math.Sqrt(normB)
	if norm == 0 {
		return 0.0
	}
	return dot / norm
}

File: rust/src/storage/vector_store.rs
//! Qdrant vector store for semantic similarity search via REST API.
//!
//! Uses HTTP calls to Qdrant REST API. Manages two collections:
//!   - memory_text: sentence-transformer embeddings of memory content
//!   - memory_visual: CLIP embeddings of scene descriptions

use log::debug;
use reqwest::Client;
use reqwest::StatusCode;
use serde::{Deserialize, Serialize};
use uuid::Uuid;

pub const TEXT_COLLECTION: &str = "memory_text";
pub const VISUAL_COLLECTION: &str = "memory_visual";

/// Qdrant REST API-backed vector store for memory embeddings.
pub struct VectorStore {
    base_url: String,
    client: Client,
}

#[derive(Debug, Serialize)]
struct UpsertRequest {
    points: Vec<PointStruct>,
}

#[derive(Debug, Serialize)]
struct PointStruct {
    id: String,
    vector: Vec<f64>,
    payload: serde_json::Value,
}

#[derive(Debug, Serialize)]
struct SearchRequest {
    vector: Vec<f64>,
    limit: usize,
    #[serde(skip_serializing_if = "Option::is_none")]
    filter: Option<serde_json::Value>,
    with_payload: bool,
}

#[derive(Debug, Deserialize)]
struct SearchResponse {
    result: Vec<SearchHit>,
}

#[derive(Debug, Deserialize)]
struct SearchHit {
    #[allow(dead_code)]
    id: serde_json::Value,
    score: f64,
    payload: Option<serde_json::Value>,
}

#[derive(Debug, Serialize)]
struct CreateCollectionRequest {
    vectors: VectorParams,
}

#[derive(Debug, Serialize)]
struct VectorParams {
    size: usize,
    distance: String,
}

#[derive(Debug, Deserialize)]
struct RetrieveResponse {
    result: Vec<RetrievePoint>,
}

#[derive(Debug, Deserialize)]
struct RetrievePoint {
    #[allow(dead_code)]
    id: serde_json::Value,
    vector: Option<Vec<f64>>,
}

#[derive(Debug, Serialize)]
struct DeleteRequest {
    filter: serde_json::Value,
}

impl VectorStore {
    pub fn new(base_url: &str) -> Self {
        Self {
            base_url: base_url.trim_end_matches('/').to_string(),
            client: Client::new(),
        }
    }

    /// Initialize and ensure collections exist.
    pub async fn initialize(
        &self,
        text_dim: usize,
        visual_dim: usize,
    ) -> Result<(), reqwest::Error> {
        self.ensure_collection(TEXT_COLLECTION, text_dim).await?;
        self.ensure_collection(VISUAL_COLLECTION, visual_dim).await?;
        Ok(())
    }

    async fn ensure_collection(
        &self,
        name: &str,
        dim: usize,
    ) -> Result<(), reqwest::Error> {
        let url = format!("{}/collections/{name}", self.base_url);

        // Check if collection exists
        let resp = self.client.get(&url).send().await?;
        match resp.status() {
            StatusCode::OK => return Ok(()),
            StatusCode::NOT_FOUND => {}
            _ => {
                resp.error_for_status()?;
            }
        }

        // Create collection
        let body = CreateCollectionRequest {
            vectors: VectorParams {
                size: dim,
                distance: "Cosine".to_string(),
            },
        };
        self.client
            .put(&url)
            .json(&body)
            .send()
            .await?
            .error_for_status()?;
        Ok(())
    }

    // ── Text embeddings ──

    /// Insert or update a text embedding. Returns the point ID.
    pub async fn upsert_text_vector(
        &self,
        memory_id: &str,
        vector: Vec<f64>,
        tier: &str,
        valence: f64,
        arousal: f64,
        session_id: &str,
        created_at: &str,
    ) -> Result<String, reqwest::Error> {
        let point_id = Uuid::new_v4().to_string();
        let payload = serde_json::json!({
            "memory_id": memory_id,
            "tier": tier,
            "valence": valence,
            "arousal": arousal,
            "session_id": session_id,
            "created_at": created_at,
        });

        let body = UpsertRequest {
            points: vec![PointStruct {
                id: point_id.clone(),
                vector,
                payload,
            }],
        };

        let url = format!("{}/collections/{TEXT_COLLECTION}/points", self.base_url);
        self.client
            .put(&url)
            .json(&body)
            .send()
            .await?
            .error_for_status()?;
        Ok(point_id)
    }

    /// Search for nearest text embeddings.
    pub async fn search_text(
        &self,
        query_vector: Vec<f64>,
        limit: usize,
        tier_filter: Option<&str>,
    ) -> Result<Vec<TextSearchResult>, reqwest::Error> {
        let filter = tier_filter.map(|tier| {
            serde_json::json!({
                "must": [{"key": "tier", "match": {"value": tier}}]
            })
        });

        let body = SearchRequest {
            vector: query_vector,
            limit,
            filter,
            with_payload: true,
        };

        let url = format!("{}/collections/{TEXT_COLLECTION}/points/search", self.base_url);
        let resp: SearchResponse = self
            .client
            .post(&url)
            .json(&body)
            .send()
            .await?
            .error_for_status()?
            .json()
            .await?;

        Ok(resp
            .result
            .iter()
            .map(|hit| {
                let payload = hit.payload.as_ref();
                TextSearchResult {
                    memory_id: payload
                        .and_then(|p| p.get("memory_id"))
                        .and_then(|v| v.as_str())
                        .unwrap_or("")
                        .to_string(),
                    score: hit.score,
                    tier: payload
                        .and_then(|p| p.get("tier"))
                        .and_then(|v| v.as_str())
                        .unwrap_or("hot")
                        .to_string(),
                    valence: payload
                        .and_then(|p| p.get("valence"))
                        .and_then(|v| v.as_f64())
                        .unwrap_or(0.0),
                    arousal: payload
                        .and_then(|p| p.get("arousal"))
                        .and_then(|v| v.as_f64())
                        .unwrap_or(0.0),
                }
            })
            .collect())
    }

    // ── Visual embeddings ──

    /// Insert or update a visual (CLIP) embedding. Returns the point ID.
    pub async fn upsert_visual_vector(
        &self,
        memory_id: &str,
        vector: Vec<f64>,
        session_id: &str,
        created_at: &str,
    ) -> Result<String, reqwest::Error> {
        let point_id = Uuid::new_v4().to_string();
        let payload = serde_json::json!({
            "memory_id": memory_id,
            "session_id": session_id,
            "created_at": created_at,
        });

        let body = UpsertRequest {
            points: vec![PointStruct {
                id: point_id.clone(),
                vector,
                payload,
            }],
        };

        let url = format!("{}/collections/{VISUAL_COLLECTION}/points", self.base_url);
        self.client
            .put(&url)
            .json(&body)
            .send()
            .await?
            .error_for_status()?;
        Ok(point_id)
    }

    /// Search for nearest visual embeddings.
    pub async fn search_visual(
        &self,
        query_vector: Vec<f64>,
        limit: usize,
    ) -> Result<Vec<VisualSearchResult>, reqwest::Error> {
        let body = SearchRequest {
            vector: query_vector,
            limit,
            filter: None,
            with_payload: true,
        };

        let url = format!(
            "{}/collections/{VISUAL_COLLECTION}/points/search",
            self.base_url
        );
        let resp: SearchResponse = self
            .client
            .post(&url)
            .json(&body)
            .send()
            .await?
            .error_for_status()?
            .json()
            .await?;

        Ok(resp
            .result
            .iter()
            .map(|hit| {
                let payload = hit.payload.as_ref();
                VisualSearchResult {
                    memory_id: payload
                        .and_then(|p| p.get("memory_id"))
                        .and_then(|v| v.as_str())
                        .unwrap_or("")
                        .to_string(),
                    score: hit.score,
                }
            })
            .collect())
    }

    /// Compute cosine similarity between two points in the text collection.
    /// Used by dream explorer (A3) for cross-session similarity checks.
    pub async fn similarity(
        &self,
        point_id_a: &str,
        point_id_b: &str,
    ) -> Result<Option<f64>, reqwest::Error> {
        let url = format!(
            "{}/collections/{TEXT_COLLECTION}/points",
            self.base_url
        );

        let body = serde_json::json!({
            "ids": [point_id_a, point_id_b],
            "with_vector": true
        });

        let resp: RetrieveResponse = self
            .client
            .post(&url)
            .json(&body)
            .send()
            .await?
            .error_for_status()?
            .json()
            .await?;

        if resp.result.len() < 2 {
            return Ok(None);
        }

        let a = match &resp.result[0].vector {
            Some(v) => v,
            None => return Ok(None),
        };
        let b = match &resp.result[1].vector {
            Some(v) => v,
            None => return Ok(None),
        };

        Ok(Some(cosine_similarity(a, b)))
    }

    /// Delete all points for a given memory_id from a collection.
    pub async fn delete_point(
        &self,
        collection: &str,
        memory_id: &str,
    ) -> Result<(), reqwest::Error> {
        let url = format!(
            "{}/collections/{collection}/points/delete",
            self.base_url
        );

        let body = DeleteRequest {
            filter: serde_json::json!({
                "must": [{"key": "memory_id", "match": {"value": memory_id}}]
            }),
        };

        self.client
            .post(&url)
            .json(&body)
            .send()
            .await?
            .error_for_status()?;
        debug!("Deleted points for memory_id={memory_id} from {collection}");
        Ok(())
    }
}

/// Compute cosine similarity between two vectors.
///
/// Returns 0.0 if vectors are empty or have different lengths.
pub fn cosine_similarity(a: &[f64], b: &[f64]) -> f64 {
    if a.len() != b.len() || a.is_empty() {
        return 0.0;
    }
    let dot: f64 = a.iter().zip(b.iter()).map(|(x, y)| x * y).sum();
    let norm_a: f64 = a.iter().map(|x| x * x).sum::<f64>().sqrt();
    let norm_b: f64 = b.iter().map(|x| x * x).sum::<f64>().sqrt();
    let norm = norm_a * norm_b;
    if norm > 0.0 {
        dot / norm
    } else {
        0.0
    }
}

/// Result from text embedding search.
#[derive(Debug, Clone)]
pub struct TextSearchResult {
    pub memory_id: String,
    pub score: f64,
    pub tier: String,
    pub valence: f64,
    pub arousal: f64,
}

/// Result from visual embedding search.
#[derive(Debug, Clone)]
pub struct VisualSearchResult {
    pub memory_id: String,
    pub score: f64,
}
